src/simulators.rs

```diff
@@ -1,61 +1,192 @@
+            resolution: resolution,
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn tick(&mut self) -> bool {
+        // TODO(irfansharif): Resolution mismatch; no possibility of generating multiple packets.
+        if self.ticker == 0 {
+            self.statistics.packets_generated += 1;
+            self.ticker = self.generator.next_event(self.resolution);
+            return true;
+        }
+
+        self.ticker -= 1;
+        if self.ticker == 0 {
+            self.statistics.packets_generated += 1;
+            self.ticker = self.generator.next_event(self.resolution);
+            true
+        } else {
+            false
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct ServerStatistics {
+    pub packets_processed: u32,
+    pub packets_dropped: u32,
+    pub idle_count: u32,
+}
+
+impl ServerStatistics {
+    fn new() -> ServerStatistics {
+        ServerStatistics {
+            packets_processed: 0,
+            packets_dropped: 0,
+            idle_count: 0,
+        }
+    }
+}
+
+// Server stores packets in a queue and processes them.
+#[allow(dead_code)]
+pub struct Server<G: Generator> {
+    queue: VecDeque<Packet>,
+    buffer_limit: Option<usize>,
+    // TODO(irfansharif): Comment.
+    ticker: u32,
+    resolution: f64,
+    generator: G,
+    pub statistics: ServerStatistics,
+}
+
+impl<G: Generator> Server<G> {
+    #[allow(dead_code)]
+    pub fn new(generator: G, resolution: f64, buffer_limit: Option<usize>) -> Server<G> {
+        Server {
+            queue: VecDeque::new(),
+            buffer_limit: buffer_limit,
             resolution: resolution,
+            ticker: generator.next_event(resolution),
+            statistics: ServerStatistics::new(),
+            generator: generator,
         }
     }
 
-    // next_packet takes the current time and uses it to determine whether to generate a packet or
-    // not. If a packet is generated, the last_gen time and till_next times are both updated.
     #[allow(dead_code)]
-    fn next_packet(&mut self, time: u32) -> Option<Packet> {
-        if time == self.next_gen {
-            self.next_gen = time + self.generator.next_event(self.resolution);
-            Some(Packet(time))
+    pub fn enqueue(&mut self, packet: Packet) {
+        match self.buffer_limit {
+            Some(limit) => {
+                if self.queue.len() < limit {
+                    self.queue.push_back(packet);
+                } else {
+                    self.statistics.packets_dropped += 1
+                }
+            }
+            // Infinite queue.
+            None => {
+                self.queue.push_back(packet);
+            }
+        }
+    }
+
+    // TODO(irfansharif): Comment.
+    #[allow(dead_code)]
+    pub fn tick(&mut self) -> Option<Packet> {
+        if self.ticker == 0 {
+            self.ticker = self.generator.next_event(self.resolution);
+            let packet = self.queue.pop_front();
+            match packet {
```

Comment:
Not sure if this is the correct behaviour: if the queue is empty, the ticker shouldn't start since there is technically nothing to "process" at that time... it looks like (correct me if I'm wrong) the ticker starts regardless of whether or not there is actually a packet in the queue.

Reply:
In the case that the queue is empty (nothing should be ticking and therefore the server is idle), we still could be stuck in the decrement loop from ticker being restarted by the generator, in which case the idle time is not incrementing and we are not actively for the next packet.

Reply:
Explained else where, added a comment explaining this pattern:
```rust
// The server operates via polling essentially, we process an entire packet each time the
// ticker returns to zero, if any. If the queue is empty this means the server is idle for that
// time duration and is recorded so internally.
```